### PR TITLE
[CS598] Added GBSG Dataset + Example running notebook with MLP Model

### PR DESCRIPTION
**Pull request by chrisyu-uiuc**
#Dataset
PyHealth\pyhealth\datasets\__init__.py
PyHealth\pyhealth\datasets\gbsg.py
#Task
PyHealth\pyhealth\tasks\__init__.py
PyHealth\pyhealth\tasks\survival_analysis.py
#Demo Running Notebook (For referencing only, delete it if approrpiate)
PyHealth\examples\survival_analysis_mlp_gbsg.ipynb

NetID: hmyu2, jl279

Related Document : Revisit Deep Cox Mixtures For Survival Regression
https://github.com/chrisyu-uiuc/revisit-deepcoxmixtures-cs598-uiuc/blob/main/Revisit_DeepCoxMixturesForSurvivalRegression.pdf

Type of Contribution: Dataset and Task Implementation

High-Level Description: This contribution involves implementing the GBSGDataset and SurvivalAnalysisGBSG classes within the pyhealth framework. The GBSGDataset class handles the loading and management of the German Breast Cancer Study Group (GBSG) dataset, providing the necessary clinical and survival data. The SurvivalAnalysisGBSG class defines the specific task of survival analysis for this dataset, structuring the data into appropriate inputs (patient features) and outputs (time-to-event and event status) for model training and evaluation.

Files to Look At:

    PyHealth\pyhealth\datasets\__init__.py

    PyHealth\pyhealth\datasets\gbsg.py

    PyHealth\pyhealth\tasks\__init__.py

    PyHealth\pyhealth\tasks\survival_analysis.py

    PyHealth\examples\survival_analysis_mlp_gbsg.ipynb (for a running example)

Press to Run Colab : 
https://colab.research.google.com/drive/1BKtboOKpLKQeiYTS-ClTqFGbz5OhZn9T?usp=sharing

Description:
This project uses a StaticHazardModel, an MLP, to predict patient risk scores from features for survival analysis on the GBSG dataset.

The model is trained with the cox_loss function, which focuses on risk ranking. It uses an Adam optimizer and a learning rate scheduler for efficient training.

Hyperparameter tuning is done with Optuna and Stratified K-Fold cross-validation. It finds optimal settings for parameters like learning rate and batch size by minimizing the Integrated Brier Score (IBS).

The best model is retrained on the full dataset. Final evaluation uses metrics like Concordance Index (Ctd), time-dependent AUC, IBS, Brier Score, and Expected Calibration Error (ECE) to assess performance.

Results :
Final Evaluation Metrics (on the full dataset):
  Concordance Index (Ctd): 0.7857
  Time-dependent AUC (avg over times): 0.8279
  Integrated Brier Score (IBS): 0.1299
  Expected Calibration Error (ECE, avg over times): 0.1765
  Brier Score (avg over times): 0.1230

Hormonal Therapy Metrics:
  Ctd: 0.7867
  AUC: 0.8474
  IBS: 0.1209
  Avg ECE: 0.2090
  Avg Brier: 0.1100

    Achieve best Ctd, AUC, BS comparing our existing baselines (DSM, DCM, CPH, RSF)

### CODE REVIEW

**Comment by jhnwu3** [comment]
Closing this PR as it lacks proper labeling. Please add appropriate labels and reopen if needed.
